File: examples/agents/simple_4.py
#!/usr/bin/env python

"""
Stage 4: classify and answer in a single call.

One schema carries both the on/off topic verdict and the answer, so the model
does both jobs in one request. The refusal is printed locally when the model
reports the question is off topic.

The modules you need to install to make this work are `passpy` and `anthropic`
"""
# These demos are deliberately flat single-file scripts: the driver code sits
# at module level so the whole agent loop reads top to bottom. That makes a
# helper's parameter shadow the module-level name it is called with.
# pylint: disable=redefined-outer-name


import json

import passpy
import anthropic

MODEL = "claude-opus-5"

AGENT_SYSTEM = """
You are a helpful airline ticket sales agent.

For every user message, do two things at once:

1. Decide whether it is related to airline ticket sales: flight searches, fares
   and pricing, booking, changes and cancellations, baggage, seating, check-in,
   refunds, loyalty programs, and travel documents for a flight. Anything else
   is off topic. Report this as `on_topic`.

2. If it is on topic, answer it in `answer`. You have no access to a live
   booking system, so be clear about what the customer would need to confirm
   with the airline directly. If it is off topic, set `answer` to an empty
   string.
"""

REFUSAL = "Sorry, I only deal with airline ticket sales."

REPLY_SCHEMA = {
    "type": "object",
    "properties": {
        "on_topic": {
            "type": "boolean",
            "description": "True if the message is about airline ticket sales.",
        },
        "answer": {
            "type": "string",
            "description": "The answer if on topic, otherwise an empty string.",
        },
    },
    "required": ["on_topic", "answer"],
    "additionalProperties": False,
}


def reply(question):
    """Classify and answer a question in one call; return the text to print."""
    response = client.messages.create(
        model=MODEL,
        max_tokens=16000,
        system=AGENT_SYSTEM,
        output_config={"format": {"type": "json_schema", "schema": REPLY_SCHEMA}},
        messages=[{"role": "user", "content": question}],
    )
    text = "".join(b.text for b in response.content if b.type == "text")
    result = json.loads(text)
    return result["answer"] if result["on_topic"] else REFUSAL


store = passpy.Store()
api_key = store.get_key("keys/claude.ai")
assert api_key is not None
api_key = api_key.rstrip()
client = anthropic.Anthropic(api_key=api_key)

print("Ask me about airline tickets. Type 'exit' or 'quit' (or Ctrl-D) to stop.")
while True:
    try:
        question = input("\n> ").strip()
    except (EOFError, KeyboardInterrupt):
        print()
        break
    if not question:
        continue
    if question.lower() in ("exit", "quit"):
        break
    print(reply(question))

File: examples/airline-ticket-agent-node/simple_4.js
#!/usr/bin/env node

/*
 * Stage 3: same question loop, but scoped to airline ticket sales.
 *
 * A system prompt tells the model to only answer questions that are at least
 * somewhat related to airline ticket sales, and to politely refuse anything
 * else. Each question is still an independent request -- no history is kept.
 *
 * The module you need to install to make this work is `@anthropic-ai/sdk`.
 * The api key is read from pass(1) via `pass show keys/claude.ai`.
 */

import { execFileSync } from "node:child_process";
import readline from "node:readline";
import Anthropic from "@anthropic-ai/sdk";

const MODEL = "claude-opus-5";

const SYSTEM = `
You are an airline ticket sales agent. Only answer questions that are at
least somewhat related to airline ticket sales: flights, fares, bookings,
schedules, baggage, seating, airports, airlines, travel documents and the
like. If the question is about anything else, do not answer it; instead
reply exactly: "I can only answer questions related to airline ticket
sales."
`.trim();

const apiKey = execFileSync("pass", ["show", "keys/claude.ai"], {
  encoding: "utf8",
}).trimEnd();
const client = new Anthropic({ apiKey });

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  prompt: "\n> ",
});
rl.on("SIGINT", () => rl.close());

console.log("Ask me about airline tickets. Press Ctrl-D to stop.");
rl.prompt();

for await (const line of rl) {
  const question = line.trim();
  if (question) {
    const response = await client.messages.create({
      model: MODEL,
      max_tokens: 16000,
      system: SYSTEM,
      messages: [{ role: "user", content: question }],
    });
    for (const block of response.content) {
      if (block.type === "text") {
        console.log(block.text);
      }
    }
  }
  rl.prompt();
}
console.log();
